### PR TITLE
8332003: Clarify javadoc for MemoryLayout::offsetHandle

### PR DESCRIPTION
**Pull request by mcimadamore**
Consider this layout:

```
MemoryLayout SEQ = MemoryLayout.sequenceLayout(5,
             MemoryLayout.sequenceLayout(10, JAVA_INT));
```

And the corresponding offset handle:

```
MethodHandle offsetHandle = SEQ.offsetHandle(PathElement.sequenceLayout(), PathElement.sequenceLayout());
```

The resulting method handle takes two additional `long` indices. The implementation checks that the dynamically provided indices conform to the bound available at construction: that is, the first index must be < 5, while the second must be < 10. If this is not true, an `IndexOutOfBoundException` is thrown.

However, the javadoc for `MemoryLayout::byteOffsetHandle` doesn't claim anything in this direction. There are only some vague claims in the javadoc for `PathElement::sequenceElement()` and `PathElement::sequenceElement(long, long, long)` which make some claims on which indices are actually allowed, but the text seems more in the tone of a discussion, rather than actual normative text.

I've tweaked the javadoc for `MemoryLayout::byteOffsetHandle` to actually state that the indices will be checked against the "size" of the corresponding open path element (this is a new concept that I also have defined in the javadoc).

I also added a test for `byteOffsetHandle` as I don't think we had a test for that specifically (although this method is tested indirectly, via `MemoryLayout::varHandle`).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8332143](https://bugs.openjdk.org/browse/JDK-8332143) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8332003](https://bugs.openjdk.org/browse/JDK-8332003): Clarify javadoc for MemoryLayout::offsetHandle (**Bug** - P3)
 * [JDK-8332143](https://bugs.openjdk.org/browse/JDK-8332143): Clarify javadoc for MemoryLayout::offsetHandle (**CSR**)

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19158/head:pull/19158` \
`$ git checkout pull/19158`

Update a local copy of the PR: \
`$ git checkout pull/19158` \
`$ git pull https://git.openjdk.org/jdk.git pull/19158/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19158`

View PR using the GUI difftool: \
`$ git pr show -t 19158`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19158.diff">https://git.openjdk.org/jdk/pull/19158.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19158#issuecomment-2102843300)

### CODE REVIEW

File: test/jdk/java/foreign/TestLayoutPaths.java [hunk 1]
```diff
@@ -36,8 +36,10 @@
 import java.lang.invoke.VarHandle;
```
Comment (minborg) [inline test/jdk/java/foreign/TestLayoutPaths.java:36]: Nit: Copyright year

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 2]
```diff
@@ -609,7 +613,7 @@ public sealed interface MemoryLayout
      * offset {@code O} of the access operation is computed as follows:
      *
      * {@snippet lang = "java":
-     * O = this.offsetHandle(P).invokeExact(B, I1, I2, ... In);
+     * O = this.byteOffsetHandle(P).invokeExact(B, I1, I2, ... In);
```
Comment (minborg) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:616]: Good catch

**Comment by mcimadamore** [comment]
In an offline discussion @minborg noted that other methods in the `MemoryLayout` class do not specify in full the set exception thrown by the returned method handle/var handle:

* MemoryLayout::varHandle
* MemoryLayout::sliceHandle
* MemoryLayout::arrayElementVarHandle

The first two are missing details re. checks against the upper bound of the index parameter/coordinates. I believe this omission was because these methods already say that if the accessed offset is bigger than the segment size, then a IIOBE occurs. But there's a subtle problem with it. In the impl, even if you have a big enough memory segment to allow access at index 42, but the layout says that the max index is 41, access fails with IIOBE anyway. So I think the javadoc should be updated to contain the new text about the index checks.

The last method in the list actually doesn't have any text when it comes to the exceptions thrown by the returned handle. This is because we say that this method can be obtained by calling `collectCoordinates` on `MemoryLayout::varHandle`. Nevertheless, I think it's worth repeating what exceptions can come up (especially as doing so reveals more clearly that there's no extra checks for the very first long coordinate of the returned handle - so now the existing `apiNote` makes even more sense, I think).

File: src/java.base/share/classes/java/lang/foreign/MemoryLayout.java [hunk 3]
```diff
@@ -571,6 +571,10 @@ public sealed interface MemoryLayout
      * derived from the size of the element layout of a sequence, and
      * {@code c_1}, {@code c_2}, ... {@code c_m} are other <em>static</em> offset
      * constants (such as field offsets) which are derived from the layout path.
+     * <p>
```
Comment (PaulSandoz) [inline src/java.base/share/classes/java/lang/foreign/MemoryLayout.java:574]: Can connected back to the prior paragraph? e.g., for any given dynamic argument `x_n`, the index into the sequence associated with the n'th open path element whose size is `size_n`,`x_n` must be `>= 0` or `< size_n`

**Review by PaulSandoz — APPROVED** [review]

**Pull request closed without merging**